File: HSGoogleDrivePicker/HSGoogleDrivePicker/HSDriveManager.m
//
//  GDriveManager.m
//  gdrive
//
//  Created by Rob Jonson on 13/10/2015.
//  Copyright © 2015 HobbyistSoftware. All rights reserved.
//

#import "HSDriveManager.h"
#import "HSGIDSignInHandler.h"


static NSString *const kKeychainItemName = @"Drive API";

@interface HSDriveManager ()

@property (retain) NSString *clientId;
@property (retain) NSString *clientSecret;
@property (nonatomic, strong) GTLServiceDrive *service;

@end

@implementation HSDriveManager

- (instancetype)initWithId:(NSString*)clientId secret:(NSString*)secret
{
    self = [super init];
    if (self) {
        
        self.clientId=clientId;
        self.clientSecret=secret;
        
        // Initialize the Drive API service & load existing credentials from the keychain if available.
        self.service = [[GTLServiceDrive alloc] init];
        
        self.service.authorizer = [HSGIDSignInHandler authoriser];
        
        self.folderId=@"root";
        self.maxResults=1000;
        self.signOutLabel = @"Sign out";
        
    }
    return self;
}

#pragma mark download

-(GTMSessionFetcher*)downloadFile:(GTLDriveFile*)file toPath:(NSString*)path withCompletionHandler:(void (^)(NSError *error))handler
{
    GTMSessionFetcher *fetcher = [self.service.fetcherService fetcherWithURLString:file.downloadUrl];
    NSURL* destinationUrl=[NSURL fileURLWithPath:path];
    [fetcher setDestinationFileURL:destinationUrl];
    
    [fetcher beginFetchWithCompletionHandler:^(NSData *data, NSError *error) {
        if (error == nil) {
            // Success.
            handler( nil);
        } else {
            NSData *data = error.userInfo[@"data"];
            NSString *dError = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            NSLog(@"An error occurred: %@", error);
            handler( error);
        }
    }];
    
    return fetcher;
}

#pragma mark file listing

-(NSString*)query
{
    NSString *query=[NSString stringWithFormat:@"'%@' in parents", self.folderId];
    if (self.sharedWithMe)
    {
        query=@"sharedWithMe";
    }
    
    if (!self.showTrashed)
    {
        query=[query stringByAppendingString:@" and trashed = false"];
    }

    
    return query;
}

// Construct a query to get names and IDs of files using the Google Drive API.
- (void)fetchFilesWithCompletionHandler:(void (^)(GTLServiceTicket *ticket, GTLDriveFileList *fileList, NSError *error))handler
{
    
    self.service.shouldFetchNextPages = self.autoFetchPages;
    
    GTLQueryDrive *query = [GTLQueryDrive queryForFilesList];
    
    query.q=[self query];
    query.fields=@"files(id,kind,mimeType,name,size,iconLink)";
    
    
    query.pageSize = self.maxResults;
    
    [self.service executeQuery:query
             completionHandler:handler];
   
}



#pragma mark auth controller

-(void)updateAuthoriser
{
    self.service.authorizer = [HSGIDSignInHandler authoriser];
}


@end

File: HSGoogleDrivePicker/HSGoogleDrivePicker/HSDriveManager.swift
import GoogleAPIClientForREST

private let kKeychainItemName = "Drive API"


/// Manage interaction with Drive
/// Drive manager is declared as an NSObject so that the completion handler in DrivePicker.pick(...) can be represented in Obj-C
@objcMembers open class HSDriveManager:NSObject {
    //* called by view controller when auth changes after a sign in *
    
    
    //* runs a fetch with the current search settings. Cancels any outstanding fetch *
    
    
    /** If the download completes succesfully, then the completion handler will be called with NULL for the error
     returns GTMHTTPFetcher - this can be used to monitor download progress.
     **/
    
    
    //* if No, shows 'my files'. Default is NO *
    open var sharedWithMe = false
    //* initially 'root'. Ignored if sharedWithMe is selected *
    open var folderId:String = "root"
    //* Default is NO *
    open var showTrashed = false
    //* Default is 1000 *
    open var maxResults = 1000
    //* Default is YES. This is not recommended by Google, but I don't want to implement multi-page logic *
    open var autoFetchPages = false
    //* Default is 'Sign out'*
    open var signOutLabel:String = "Sign out"
    
    private var service: GTLRDriveService?
    
    override init() {

        // Initialize the Drive API service & load existing credentials from the keychain if available.
        service = GTLRDriveService()
        
        service?.authorizer = HSGIDSignInHandler.authoriser
    }
    
    // MARK: download
    @discardableResult
    @objc public func downloadFile(_ file: GTLRDrive_File?, toPath path: String, withCompletionHandler handler: @escaping (_ error: Error?) -> Void) -> GTMSessionFetcher? {
        
        guard let downloadURL = file?.downloadURL else {
            return nil
        }
        
        let destinationUrl = URL(fileURLWithPath: path)
        
        let fetcher = service?.fetcherService.fetcher(with: downloadURL)
        fetcher?.destinationFileURL = destinationUrl
        
        fetcher?.beginFetch(completionHandler: { data, error in
            if error == nil {
                // Success.
                handler(nil)
            } else {

                if let error = error {
                    print("An error occurred: \(error)")
                }
                if let data = (error as NSError?)?.userInfo["data"] as? Data  {
                    let dError = String(data: data, encoding: .utf8)
                    print("Error data: \(String(describing: dError))")
                }
                handler(error)
            }
        })
        
        return fetcher
    }
    
    // MARK: file listing
    var filterQuery:String {
        var query = "'\(folderId)' in parents"
        if sharedWithMe {
            query = "sharedWithMe"
        }
        
        if !showTrashed {
            query = query + (" and trashed = false")
        }

        return query
    }
    
    // Construct a query to get names and IDs of files using the Google Drive API.
    func fetchFiles(withCompletionHandler handler: @escaping GTLRServiceCompletionHandler) {
        
        service?.shouldFetchNextPages = autoFetchPages
        
        let query = GTLRDriveQuery_FilesList.query()
        
        query.q = self.filterQuery
        query.fields = "files(id,kind,mimeType,name,size,iconLink)"
        
        
        query.pageSize = maxResults
        
        service?.executeQuery(query, completionHandler: handler)
        
    }
    
    // MARK: auth controller
    func updateAuthoriser() {
        service?.authorizer = HSGIDSignInHandler.authoriser
    }
}
